### ISSUE TITLE
Do the different quickstats R packages want to coordinate future development?

### ISSUE THREAD

**Issue by lindblb** [ropensci/rnassqs#22]

**Comment by econpotter** [comment]
Hi @lindblb, there are several R packages all come out in the past year or so that access the quick stats data. `rnassqs` was accepted to CRAN back in May I think, and I [first posted a commit to github back in 2015](https://github.com/ropensci/rnassqs/commits/master?after=3b4265faa461bb835924796f3d4b86b141f7abd3+139), though it took a few years to really organize it into a package. AFAIK yours is the only one that provides the mapping aspect though.

Another complete package is @rdinter's [usdarnass](https://github.com/rdinter/usdarnass). There are few others that seem incomplete lying around on github. It may make sense to merge these all together if either of either of you feel it would be worth it. We could issue a new CRAN and ropensci version that incorporates everything. On the other hand if it makes sense to continue to develop separately we can do that as well.

As for future plans, I'm not planning on adding any mapping features, so `tidyUSDA` provides a useful feature set there. There are some other ideas I've posted as issues for the `rnassqs` package if I get around to them. 

I'm reopening this for continued conversation for now.

**Comment by rdinter** [comment]
I can only speak on the [usdarnass](https://github.com/rdinter/usdarnass) package's intent and potential development. The goal was to create a package for CRAN which largely mimicked what the web interface of Quick Stats could do. For example, while the API allows you to enter a county fips code for a query (the county_ansi category which is annoyingly the three digit county fips in a given state and not the typically used 5 digit fips code) the web interface does not. I intentionally omitted the ability to use county_ansi in a query because that would not be consistent with what the web interface allows. In that vein, there are only a few issues still left to incorporate in that package to match what the web interface can do and if/when those are fixed I don't foresee that package having further development unless NASS makes large changes to their web interface.

I would be more than happy to merge any of the aspects from usdarnass that could be helpful, although I don't know what could be helpful. That's more of a question for @potterzot and what he feels are aspects from usdarnass that would be good to incorporate. Candidly, I use RStudio frequently and enjoy that usdarnass has each of the categories in a Quick Stats query as an argument in the function call. That helps me because I might know that I want to get corn acres for Ohio in 2017 but I'll forget the category for corn (is it com_desc? commodity_cat? commodity_desc?) as well as what measures acres (statisticcat_desc? statisticdesc_cat? stat_cat?). Using tab after typing the function name allows me to cycle through all the function arguments and it'll provide me a short description of what each argument entails so I can identify what I need fairly quickly instead of going on to [https://quickstats.nass.usda.gov/api](https://quickstats.nass.usda.gov/api). With how [rnassqs](https://github.com/ropensci/rnassqs) is set up, I'm not sure if that is something that should/could be incorporated with how the package is currently designed as I have not spent enough time reading the source code. There's also pluses and minuses to not having categories as arguments in a function call. Like I mentioned above, it's impossible to add in a `county_ansi = 001` to a query in usdarnass but from what I can tell that's straightforward to do in rnassqs. It's also possible to add in `county_fips = 001` in rnassqs which will produce an error and the user might not know why there is an error.

**Comment by econpotter** [comment]
@rdinter thanks, good points about the differences in approach, I had forgotten about that. I don't have a strong opinion either way as far as merging, and it may be helpful for there to be different packages that meet slightly different user needs. I wouldn't have much time to work on a merge anyway for now. If it seems useful in the future I'm certainly open to it, and to continue sharing about ideas.

**Comment by econpotter** [comment]
I also changed the issue title to reflect the conversation topic, hope that's okay.

**Comment by rdinter** [comment]
I will take a look at rnassqs more to better understand the structure of the package. There's many ways to skin a cat and I suspect the functions to get to API calls for rnassqs are not in the same fashion as usdarnass. So it might not be possible to add in documentation for each argument without making major overhauls, but I'll look to see if I can help.

**Comment by rdinter** [comment]
Having looked through `rnassqs` I think I have a good understanding of the structure of the package in terms of how the various functions interact with each other. I particularly like how `nassqs_GET` is structured and reduces duplication of the API call for the Quick Stats functions (count, data, parameters).

A few things came to mind as I was able to cobble together a [pull request](https://github.com/ropensci/rnassqs/pull/24) that added in documentation for the major parameters in Quick Stats to `nassqs` that didn't break the current workflow of someone using `rnassqs`:

- The `nassqs_params()` function or some variant of that might be a good place to insert documentation for all of the parameters in Quick Stats that would return a list. Some advantages to this would be the ability to hardcode all of the parameters to safeguard typos and a user on RStudio could use tab to cycle through all the potential parameters. Currently, one usually passes a manually created `list()` of parameters. I think it is a good idea to push users towards an alternate version of something like `nassqs_params()` that would only accept arguments that are on Quick Stats. I think that would cut down on user frustration of typos in their parameter names. I'm still trying to think of a way to correct for typos of values like "CRON" instead of "CORN" but nothing comes to mind.
- Should the definitions of each parameter in `nassqs_params()` be different from what Quick Stats has provided in their API documentation? I lean towards keeping the definitions that NASS currently has in the [API](https://quickstats.nass.usda.gov/api).
- The default for format in `nassqs_GET` is JSON, however IIRC the download size is generally smallest for CSV from some local tests that I did. It's not a big issue probably something worth double checking.
- County FIPS is an obvious argument to be implemented in this package! It's frustrating that `county_code` is not was most people would regard as a FIPS code. A simple concatenation of `state_fips_code` and `county_code` creates a FIPS so it should be straightforward to reverse this in making API calls (ie `county_fips` as an argument that gets passed along to `state_fips_code` and `county_code` in an API call)
